Stackoverflow when using the eval target

[Stumbled upon](https://github.com/nadako/hxtsdgen/issues/5) this issue while using [hxtsdgen](https://github.com/nadako/hxtsdgen) with Haxe `4.0.0-rc.2+92c7833`.

Here's the stacktrace:

```
Fatal error: exception Stack overflow
Raised by primitive operation at file "hashtbl.ml", line 163, characters 11-24
Called from file "hashtbl.ml", line 169, characters 18-34
Called from file "_build/src/macro/eval/evalHash.ml", line 26, characters 1-32
Called from file "_build/src/macro/eval/evalEncode.ml", line 125, characters 36-44
Called from file "list.ml", line 55, characters 20-23
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "list.ml", line 55, characters 32-39
Called from file "_build/src/macro/eval/evalEncode.ml", line 125, characters 12-50
Called from file "_build/src/macro/eval/evalEmitter.ml", line 441, characters 9-17
Called from file "_build/src/macro/eval/evalEmitter.ml", line 79, characters 23-31
Called from file "_build/src/macro/eval/evalEmitter.ml", line 245, characters 7-18
Called from file "_build/src/macro/eval/evalEmitter.ml", line 744, characters 9-17
Called from file "_build/src/macro/eval/evalEmitter.ml", line 744, characters 9-17
Called from file "_build/src/macro/eval/evalEmitter.ml", line 744, characters 9-17
Called from file "_build/src/macro/eval/evalEmitter.ml", line 744, characters 9-17
Called from file "_build/src/macro/eval/evalEmitter.ml", line 744, characters 9-17
Called from file "_build/src/macro/eval/evalEmitter.ml", line 744, characters 9-17
Called from file "_build/src/macro/eval/evalEmitter.ml", line 744, characters 9-17
Called from file "_build/src/macro/eval/evalEmitter.ml", line 744, characters 9-17
Called from file "_build/src/macro/eval/evalEmitter.ml", line 744, characters 9-17
...
```

Two ways to reproduce that I know of:

1) Clone [hxtsdgen](https://github.com/nadako/hxtsdgen), install the `hxnodejs` haxelib and then run `make`. You'll see it will fail at the `function.text` test case;
1) You can clone [this](https://github.com/haxe-boilerplate/pwa-ts-haxe-boilerplate-take-2), install the necessary Haxelibs and then, at the root of the project dir, run:`haxe build-client.hx`. `hxtsdgen` fails when trying to generate the TS type definitions for the return type of the `call()`method (`tink.Future<Result>`), [here](https://github.com/haxe-boilerplate/pwa-ts-haxe-boilerplate-take-2/blob/master/src/hx/client/TinkProxy.hx#L15).

To see the OCaml stacktrace, export the `OCAMLRUNPARAM=b` env var.

Currently this is a blocker for using hxtsdgen with Haxe 4.

I'm not getting a stack overflow, it seems to just fail:

```
C:\GitHub\hxtsdgen>make
haxe test.hxml
node test
Running test case `basic-types.txt`...
Running test case `callback.txt`...
Running test case `ctors.txt`...
Running test case `doc.txt`...
Running test case `eithertype.txt`...
Running test case `empty.txt`...
Running test case `expose.txt`...
Running test case `function.txt`...
Haxe compilation failed!
Running test case `generic.txt`...
Running test case `readonly.txt`...
Running test case `structs.txt`...
Haxe compilation failed!
Result: 11. Failed: 2.
make: *** [Makefile:3: all] Error 1
```

Which makes this tricky to investigate. Somebody likely has to reduce this...

@Simn 

My output does show a `Stack overflow` output before the `Haxe compilation failed!` appears in the following line:

```
 ~/workspace/code/hxtsdgen (master●)$ make                                             [2.5.0]
haxe test.hxml
node test
Running test case `basic-types.txt`...
Running test case `callback.txt`...
Running test case `ctors.txt`...
Running test case `doc.txt`...
Running test case `eithertype.txt`...
Running test case `empty.txt`...
Running test case `expose.txt`...
Running test case `function.txt`...
Stack overflow
Haxe compilation failed!
Running test case `generic.txt`...
Running test case `readonly.txt`...
Running test case `structs.txt`...
Stack overflow
Haxe compilation failed!
Result: 11. Failed: 2.
Makefile:2: recipe for target 'all' failed
make: *** [all] Error 1
```

Using Haxe `4.0.0-rc.2+77068e1` on Linux Mint.

To see the OCaml stacktrace, export the `OCAMLRUNPARAM=b` env var.

Here's the full output with the OCaml stracktrace: https://www.pastiebin.com/5ce2547458631.

--- a/src/macro/eval/evalEmitter.ml
+++ b/src/macro/eval/evalEmitter.ml
@@ -210,7 +210,7 @@ let emit_try exec catches env =
 		List.iter (fun (_,path,_) -> Hashtbl.remove eval.caught_types path) catches
 	in
 	let v = try
-		let v = exec env in
+		let v = handle_stack_overflow eval (fun() -> exec env) in
 		restore();
 		v
 	with RunTimeException(v,_,_) as exc ->

--- a/src/macro/eval/evalExceptions.ml
+++ b/src/macro/eval/evalExceptions.ml
@@ -122,17 +122,18 @@ let build_exception_stack ctx env =
 		{pfile = rev_hash env.env_info.pfile;pmin = env.env_leave_pmin; pmax = env.env_leave_pmax},env.env_info.kind
 	) d
 
+let handle_stack_overflow eval f =
+	try f()
+	with Stack_overflow ->
+		raise (RunTimeException (EvalString.create_unknown "Stack overflow", call_stack eval, null_pos))
+
 let catch_exceptions ctx ?(final=(fun() -> ())) f p =
 	let prev = !get_ctx_ref in
 	select ctx;
 	let eval = get_eval ctx in
 	let env = eval.env in
 	let r = try
-		let v =
-			try f()
-			with Stack_overflow as e ->
-				raise (RunTimeException (EvalString.create_unknown "Stack overflow", call_stack eval, null_pos))
-		in
+		let v = handle_stack_overflow eval f in
 		get_ctx_ref := prev;
 		final();
 		Some v

--- a/src/macro/eval/evalExceptions.ml
+++ b/src/macro/eval/evalExceptions.ml
@@ -128,7 +128,11 @@ let catch_exceptions ctx ?(final=(fun() -> ())) f p =
 	let eval = get_eval ctx in
 	let env = eval.env in
 	let r = try
-		let v = f() in
+		let v =
+			try f()
+			with Stack_overflow as e ->
+				raise (RunTimeException (EvalString.create_unknown "Stack overflow", call_stack eval, null_pos))
+		in
 		get_ctx_ref := prev;
 		final();
 		Some v

--- a/src/macro/eval/evalExceptions.ml
+++ b/src/macro/eval/evalExceptions.ml
@@ -124,8 +124,7 @@ let build_exception_stack ctx env =
 
 let handle_stack_overflow eval f =
 	try f()
-	with Stack_overflow ->
-		raise (RunTimeException (EvalString.create_unknown "Stack overflow", call_stack eval, null_pos))
+	with Stack_overflow -> exc_string "Stack overflow"
 
 let catch_exceptions ctx ?(final=(fun() -> ())) f p =
 	let prev = !get_ctx_ref in

